python/ray/autoscaler/node_provider.py

```diff
@@ -206,6 +227,43 @@ def internal_ip(self, node_id):
         """Returns the internal ip (Ray ip) of the given node."""
         raise NotImplementedError
 
+    def get_node_id(self, ip_address, use_internal_ip=False) -> str:
+        """Returns the node_id given an IP address.
+
+        Assumes ip-address is unique per node.
+
+        Args:
+            ip_address (str): Address of node.
+            use_internal_ip (bool): Whether the ip address is
+                public or private.
+
+        Raises:
+            ValueError if not found.
+        """
+
+        def find_node_id():
+            if use_internal_ip:
+                return self._internal_ip_cache.get(ip_address)
+            else:
+                return self._external_ip_cache.get(ip_address)
+
+        if not find_node_id():
+            all_nodes = self.non_terminated_nodes({})
+            for node_id in all_nodes:
+                self._external_ip_cache[self.external_ip(node_id)] = node_id
```

Comment:
Can we put an if statement here and only do the call that's necessary? 

`self.external_ip` and `self.internal_ip` sometimes make API calls (for example on AWS).

Reply:
Are you sure that's preferable?

For our 3 major cloud providers, `self.non_terminated_nodes({})` already populates the node cache, which means external_ip will (almost) never be making an API call.

Reply:
Actually there's a bigger problem here, which is that this breaks on node providers without a concept of external node provider (like k8s and staroid), since the unused external call will throw. 

Also, some people have projects with custom node providers which aren't cached, so if it's not too much trouble, it would be nice to support it.

Reply:
pushed an update. 

FYI we actually override this for kubernetes, and custom node providers without external IPs need to provide a custom implementation anyways because use_internal_ip=False will not work.